src/freertos_drivers/arduino/Esp32HardwareCanAdapter.hxx

```diff
@@ -0,0 +1,209 @@
+ *
+ * @author Mike Dunston
+ * @date 19 January 2019
+ */
+
+// This include is exclusive against freertos_drivers/arduino/Esp32HardwareCan.hxx
+#ifndef _FREERTOS_DRIVERS_ARDUINO_ESP32HWCAN_HXX_
+#define _FREERTOS_DRIVERS_ARDUINO_ESP32HWCAN_HXX_
+
+#include "freertos_drivers/arduino/Can.hxx"
+#include <driver/can.h>
+#include <driver/gpio.h>
+
+class Esp32Can : public Can
+{
+public:
+    Esp32Can(const char *name, gpio_num_t rxPin, gpio_num_t txPin) : Can(name)
+    {
+        can_timing_config_t can_timing_config = CAN_TIMING_CONFIG_250KBITS();
+        can_filter_config_t can_filter_config = CAN_FILTER_CONFIG_ACCEPT_ALL();
+        // Note: not using the CAN_GENERAL_CONFIG_DEFAULT macro due to a missing
+        // cast for CAN_IO_UNUSED.
+        can_general_config_t can_general_config = {
+            .mode = CAN_MODE_NORMAL,
+            .tx_io = txPin,
+            .rx_io = rxPin,
+            .clkout_io = (gpio_num_t)CAN_IO_UNUSED,
+            .bus_off_io = (gpio_num_t)CAN_IO_UNUSED,
+            .tx_queue_len = 5,
+            .rx_queue_len = 5,
+            .alerts_enabled = CAN_ALERT_NONE,
+            .clkout_divider = 0
+        };
+        ESP_ERROR_CHECK(can_driver_install(&can_general_config,
+                &can_timing_config, &can_filter_config));
+        xTaskCreatePinnedToCore(rx_task, name, OPENMRN_STACK_SIZE, this,
+            OPENMRN_TASK_PRIORITY, nullptr, tskNO_AFFINITY);
+    }
+
+    ~Esp32Can()
+    {
+    }
+
+    /**< function to enable device */
+    virtual void enable()
+    {
+        esp_err_t start_result = can_start();
+        ESP_ERROR_CHECK(start_result);
+    }
+
+    /**< function to disable device */
+    virtual void disable()
+    {
+        esp_err_t stop_result = can_stop();
+        ESP_ERROR_CHECK(stop_result);
+    }
+protected:
+    /**< function to try and transmit a message */
+    virtual void tx_msg()
+    {
+        /* see if we can send anything out */
+        struct can_frame *can_frame;
+
+        size_t msg_count = txBuf->data_read_pointer(&can_frame);
+        unsigned txCount;
+        for (txCount = 0; txCount < msg_count; ++txCount, ++can_frame)
+        {
+            can_message_t msg;
+            msg.identifier = can_frame->can_id;
+            msg.data_length_code = can_frame->can_dlc;
+            msg.data[0] = can_frame->data[0];
+            msg.data[1] = can_frame->data[1];
+            msg.data[2] = can_frame->data[2];
+            msg.data[3] = can_frame->data[3];
+            msg.data[4] = can_frame->data[4];
+            msg.data[5] = can_frame->data[5];
+            msg.data[6] = can_frame->data[6];
+            msg.data[7] = can_frame->data[7];
+            if(can_frame->can_eff)
+            {
+                msg.flags |= CAN_MSG_FLAG_EXTD;
+            }
+            if(can_frame->can_rtr)
+            {
+                msg.flags |= CAN_MSG_FLAG_RTR;
+            }
+            esp_err_t tx_result = can_transmit(&msg, pdMS_TO_TICKS(10));
```

Comment:
this is the only API at this time but I'll add a queue and send it to the queue for TX on a diff task, similar to RX. This code will instead just be a "convert and queue" and if it sends to the queue it will be considered "sent" and the TX task will need to handle retries for sending to "can_transmit" if there is a failure in queuing for TX.

also TX / RX buffers are set to 5 packets currently, is this sufficient?